app/src/main/java/com/android/streetworkapp/model/workout/WorkoutRepositoryFirestore.kt

```diff
@@ -0,0 +1,195 @@
+package com.android.streetworkapp.model.workout
+
+import android.util.Log
+import com.google.firebase.firestore.DocumentSnapshot
+import com.google.firebase.firestore.FirebaseFirestore
+import kotlinx.coroutines.tasks.await
+
+class WorkoutRepositoryFirestore(private val db: FirebaseFirestore) : WorkoutRepository {
+
+  companion object {
+    private const val COLLECTION_PATH = "workoutData"
+    private const val ERROR_UID_EMPTY = "The user UID must not be empty."
+    private const val ERROR_TAG = "FirestoreError"
+    private const val ERROR_SESSION_ID_EMPTY = "The session ID must not be empty."
+    private const val WORKOUT_SESSIONS = "workoutSessions"
+  }
+
+  /**
+   * Retrieves the WorkoutData for a user or initializes a new one if none exists.
+   *
+   * @param uid The UID of the user.
+   */
+  override suspend fun getOrAddWorkoutData(uid: String): WorkoutData {
+    require(uid.isNotEmpty()) { ERROR_UID_EMPTY }
+    return try {
+      val document = db.collection(COLLECTION_PATH).document(uid).get().await()
+      if (!document.exists()) {
+        val workoutData = WorkoutData(uid, emptyList())
+        db.collection(COLLECTION_PATH).document(uid).set(workoutData).await()
+        workoutData
+      } else {
+        documentToWorkoutData(document)
+      }
+    } catch (e: Exception) {
+      Log.e(ERROR_TAG, "Error getting or creating WorkoutData for user UID: ${e.message}")
+      WorkoutData(uid, emptyList())
+    }
+  }
+
+  /**
+   * Adds or updates a WorkoutSession in the user's WorkoutData.
+   *
+   * @param uid The UID of the user (also the WorkoutData ID).
+   * @param workoutSession The WorkoutSession to add or update.
+   */
+  override suspend fun addOrUpdateWorkoutSession(uid: String, workoutSession: WorkoutSession) {
+    require(uid.isNotEmpty()) { ERROR_UID_EMPTY }
+    require(workoutSession.sessionId.isNotEmpty()) { ERROR_SESSION_ID_EMPTY }
+    try {
+      val workoutData = getOrAddWorkoutData(uid)
+      val updatedSessions =
+          workoutData.workoutSessions.toMutableList().apply {
+            removeAll { it.sessionId == workoutSession.sessionId }
+            add(workoutSession)
+          }
+      db.collection(COLLECTION_PATH).document(uid).update(WORKOUT_SESSIONS, updatedSessions).await()
+    } catch (e: Exception) {
+      Log.e(ERROR_TAG, "Error adding or updating WorkoutSession: ${e.message}")
+    }
+  }
+
+  /**
+   * Updates specific details of a WorkoutSession.
+   *
+   * @param uid The UID of the user (also the WorkoutData ID).
+   * @param sessionId The ID of the session to update.
+   * @param exercises A list of exercises to add or update.
+   * @param endTime The end time of the session.
+   */
+  override suspend fun updateWorkoutSessionDetails(
```

Comment:
Why only allowing to update the list of exercises and endTime ?
I think we would at least to be able to update the winner aswell as we only know the winner at the end of the session.


```
    val startTime: Long,
    val endTime: Long,
    val sessionType: SessionType,
    val participants: List<String> = emptyList(),
    val exercises: List<Exercise> = emptyList(),
    val winner: String? = null
```

Reply:
Yep changes done